<a href="https://colab.research.google.com/github/Ayushsrinivas7/PCA_Testing/blob/main/Task_PCA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### `Task` How dimensionality reduction using Principal Component Analysis (PCA) on the Wine Quality dataset contributes to improving the classification accuracy and efficiency of wine type.

Note : Use KNN for Classification.

Data Link :  [Wine Data](https://docs.google.com/spreadsheets/d/e/2PACX-1vQDVwxneOKOaJL13QMhkAhYrgWlH1tICY7RacUnj_lL8m9uUWaaUf3p7bScNyh_D2Rvt7nc1q11adSy/pub?gid=647503637&single=true&output=csv)

In [21]:
# Data Loading
import pandas as pd
wine_data_path = "https://docs.google.com/spreadsheets/d/e/2PACX-1vQDVwxneOKOaJL13QMhkAhYrgWlH1tICY7RacUnj_lL8m9uUWaaUf3p7bScNyh_D2Rvt7nc1q11adSy/pub?gid=647503637&single=true&output=csv"
wine = pd.read_csv(wine_data_path)
wine.head(1)

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0  white            7.0              0.27         0.36            20.7   

   chlorides  free sulfur dioxide  total sulfur dioxide  density   pH  \
0      0.045                 45.0                 170.0    1.001  3.0   

   sulphates  alcohol  quality  
0       0.45      8.8        6

In [7]:
#  step 1 should always be the checking the data with .info
wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   type                  6497 non-null   object 
 1   fixed acidity         6487 non-null   float64
 2   volatile acidity      6489 non-null   float64
 3   citric acid           6494 non-null   float64
 4   residual sugar        6495 non-null   float64
 5   chlorides             6495 non-null   float64
 6   free sulfur dioxide   6497 non-null   float64
 7   total sulfur dioxide  6497 non-null   float64
 8   density               6497 non-null   float64
 9   pH                    6488 non-null   float64
 10  sulphates             6493 non-null   float64
 11  alcohol               6497 non-null   float64
 12  quality               6497 non-null   int64  
dtypes: float64(11), int64(1), object(1)
memory usage: 660.0+ KB


# we can see there are null values <br>
# we have to remove them

In [9]:
wine.isna().sum()

type                     0
fixed acidity           10
volatile acidity         8
citric acid              3
residual sugar           2
chlorides                2
free sulfur dioxide      0
total sulfur dioxide     0
density                  0
pH                       9
sulphates                4
alcohol                  0
quality                  0
dtype: int64

Now also check for the duplicates

In [11]:
wine.duplicated().sum()

1168

# Now lets drop the rows that has na and also the rows that are duplicate

In [25]:
# Load the Wine Quality dataset
wine_data = pd.read_csv(wine_data_path)
print("Wine Data Shape :", wine_data.shape)
# Drop rows with missing values
wine_data = wine_data.dropna()

# Dropping Duplicates rows
wine_data.drop_duplicates(inplace=True)
print("Wine Data Shape (After Dropping-) :", wine_data.shape)
wine = wine_data
print( wine.shape)

Wine Data Shape : (6497, 13)
Wine Data Shape (After Dropping-) : (5295, 13)
(5295, 13)


In [26]:

# preparing the data for training
from sklearn.model_selection import train_test_split

x = wine.iloc[: , 1:]
y = wine.iloc[: , 0 ]
xtrain , xtest , ytrain , ytest = train_test_split( x , y , test_size = 0.2  )
print( x.shape )
print( xtrain.shape )
print( xtest.shape )
print( ytrain.shape )
print( ytest.shape )



(5295, 12)
(4236, 12)
(1059, 12)
(4236,)
(1059,)


In [28]:
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
# scaling the data
scaler = StandardScaler()
xtrain = scaler.fit_transform(xtrain)
xtest = scaler.fit_transform( xtest )

# just see how the data is performing on the knn with all features

pca = PCA( n_components = None )
xtrain_trf = pca.fit_transform( xtrain )
xtest_trf = pca.fit_transform( xtest )
knn = KNeighborsClassifier()
knn.fit( xtrain_trf , ytrain )
ypred = knn.predict( xtest_trf)

# lets see the accuracy
from sklearn.metrics import accuracy_score
print(  accuracy_score( ypred , ytest ) )







0.9754485363550519


In [29]:
# getting the optimum number of PCAs or newfeatures

feat_siz = wine.shape[1] - 1
print(feat_siz)
pca = PCA( n_components = None  )
xtrain_trf = pca.fit_transform( xtrain )
xtest_trf = pca.transform( xtest )
percent = 0
cnt = 0
arr = pca.explained_variance_ratio_
for i in range( 0 , feat_siz ):
  percent =  percent + ( arr[i] * 100 )
  cnt = cnt +1

  if percent > 90 :
    break
print( cnt )


12
8


# we can see that the total features = 12 but if we take 8 features also 90 percent of the variance can be  explained

In [30]:
pca = PCA( n_components = 8  )
xtrain_trf = pca.fit_transform( xtrain )
xtest_trf = pca.transform( xtest )
knn = KNeighborsClassifier()
knn.fit( xtrain_trf , ytrain )
ypred = knn.predict( xtest_trf)

# lets see the accuracy
from sklearn.metrics import accuracy_score
print(  accuracy_score( ypred , ytest ) )



0.987724268177526


we can see that with the 12 rows we got the 97% accuracy
but with the 8 new pca features PCAs we are getting the 98%